staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go

```diff
@@ -161,3 +190,206 @@ func RecoverFromPanic(err *error) {
+	clock clock.Clock
+}
+
+// errKey tracks uniqueness based on the caller's stack and the type/message of the error
+// it is stored in both the count map and the LRU cache
+// it is removed from the count map when it gets evicted from the LRU cache
+// it is comparable via ==
+type errKey struct {
+	stack   string
+	errType reflect.Type
+	// message is the err.Error() value
+	message string
+}
+
+// errVal tracks how many times we have seen an error, and the last time we logged it
+type errVal struct {
+	count  uint64
+	logged time.Time
+}
+
+// handleErr logs the given error if it is considered new or "not recently logged"
+// currently it logs errors whenever:
+// 1. the associated errKey does not exist in d.count (see d.getVal for specifics on how this is calculated)
+// 2. the associated counter is a power of two
+// 3. the associated logged time's delta from the current time is greater than d.delta
+func (d *dedupingErrorHandler) handleErr(err error) {
+	// the operations below that do not acquire the lock do not mutate d
+
+	// we must determine our stack in this function since getStack counts frames
+	stack := d.getStack()
+	key := errKey{stack: stack, errType: reflect.TypeOf(err), message: err.Error()}
+
+	// operations after this point can mutate d
+	// we must acquire the lock before calling getVal since we cannot
+	// have concurrent reads and writes against d.count
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	// increment our counter
+	val, isNewErr := d.getVal(&key)
+	val.count++
+
+	if isNewErr {
+		// we did not find the error, so add
+		// the associated entry in the LRU cache
+		d.cache.Add(key, nil)
+	} else {
+		// we found this error, so tell the cache that we saw it
+		// cache.Get should always return nil, true
+		d.cache.Get(key)
+	}
+
+	// determine if we need to log this time
+	if isNewErr || isPowerOfTwo(val.count) || d.clock.Since(val.logged) >= d.delta {
+		val.logged = d.clock.Now()
+		d.logError(err, val.count)
+	}
+
+	// update the counter in the map after we determine if we need to log it
+	d.count[key] = val
+}
+
+// getVal returns the errVal associated with key, and if the key represents a new error.
+// If the exact key does not exist in d.count, a similar enough fuzzy match on key.message
+// will be used as a fallback (the stack and error type must always match via ==).
+// If fuzzy matching is used, the given key's message will be updated to the similar message.
+// Thus this method requires that key be passed in as a pointer.
+func (d *dedupingErrorHandler) getVal(key *errKey) (errVal, bool) {
+	val, isOldErr := d.count[*key]
+	// found direct match, use that before doing levenshtein fuzzy lookup
+	if isOldErr {
+		return val, false // return false because this is not a new error
+	}
+
+	// Do not bother doing fuzzy checks if the direct match failed and we are set to require identical strings
+	// TODO remove this once we have a useful default value for similar
+	if d.Similar >= 1 {
```

Comment:
I would just remove all of the fuzzy matching from this PR and introduce a second PR where that aspect can be focused on.